### PR TITLE
EQREL: Merge 'extend' with insertion

### PR DESCRIPTION
**Pull request by langston-barrett**
This is intended to help with performance. Previously 'extend' was
called, which adds (up to) all the tuples from the old relation to the
new relation, right before inserting all the tuples from the new
relation back into the old relation. But simply switching the order of
insertion and extend still isn't optimal - Then the old relation would
contain a copy of the new one when 'extend' is called. So instead,
combined the two operations.

Furthermore, insertion previously used the "relation" interface to the
EQREL, which may be less performant than using the "eqrel" interface as
in this commit.

See PR #2071 for discussion.

### CODE REVIEW

**Review by b-scholz — DISMISSED** [review]

**Comment by kevzhumba** [comment]
Added some tests and did some benchmarking: on the toy program which Langston ran his initial benchmarks on in [#2054](https://github.com/souffle-lang/souffle/issues/2054), I now see the following curves:

The toy program code:
`from subprocess import check_call, DEVNULL
from re import sub
from time import time

RATE = 100

PROG = """
.decl e(x: number, y: number) eqrel
.decl o(x: number)
e(0, 0).
e(x, x + 1) :-
  x < {},
  e(x, x).
o(x) :- e(0, x).
.output o(IO=stdout)
"""

print("Max,Time")
for n in range(10):
    max_val = (n + 1) * RATE

    with open("test.dl", mode="w") as f:
        f.write(str(PROG).format(max_val))
    
    begin = time()
    check_call(["souffle", "test.dl"], stdout=DEVNULL)
    end = time()
    print(f"{max_val},{int(end-begin)}")`

This is the plot for the execution time of the program with the new insertion changes:
plotNew

This is the plot for the execution time of the program under the old insertion scheme:
plotOld

We can see that there is a significant amount of speedup, but this still doesn't fix the quadratic slowdown we are seeing; see the following plot for the same program with the new insertion changes, just over more iterations: 
plotNewExtended

Also ran this change on our actual pointer analysis multiple times, and each time saw a speedup of around 1 minute from 15 minutes to 14 minutes.

**Comment by b-scholz** [comment]
That is great!

**Comment by b-scholz** [comment]
Please, could you run `clang-format -i <header/cpp-files>`?

File: src/ast2ram/seminaive/UnitTranslator.cpp [hunk 1]
```diff
@@ -195,14 +195,14 @@ Own<ram::Statement> UnitTranslator::generateMergeRelations(
     }
 
     // Predicate - insert all values
+    if (rel->getRepresentation() == RelationRepresentation::EQREL) {
+        return mk<ram::Extend>(destRelation, srcRelation);
+    }
     for (std::size_t i = 0; i < rel->getArity(); i++) {
```
[the remark this answers is not in the document]
Comment (b-scholz) [reply src/ast2ram/seminaive/UnitTranslator.cpp:201]: Yes - you are right!

**Comment by b-scholz** [comment]
There is still the issue that the domain traversals of equivalence relations are too expensive at the moment. 

For example, 
```
.decl R(x:number, y:number) eqrel 
R(1,2).
R(2,3).
R(x,y) :- A(x,y). 

.decl D(x:number)
D(x) :- R(x,x).
.output D 
```

For the rule `D(x) :- R(x,x)`, the first and second argument of atom D are identical.  
In such a situation, we should have an iterator that enumerates over the domain of the equivalence relation only. In the current implementation, we represent the find-union set as a relation and start enumerating over all pairs and filter out pairs whose first and second element are the same. This is not ideal since the expansion from a union-find set to a relation is expensive.

File: src/ram/MergeExtend.h [hunk 2]
```diff
@@ -48,15 +48,15 @@ class Extend : public BinRelationStatement {
...
+    MergeExtend* cloning() const override {
+        auto* res = new MergeExtend(second, first);
         return res;
     }
 
 protected:
     void print(std::ostream& os, int tabpos) const override {
         os << times(" ", tabpos);
-        os << "EXTEND " << getTargetRelation() << " WITH " << getSourceRelation();
+        os << "MERGE EXTEND " << getTargetRelation() << " WITH " << getSourceRelation();
```
Comment (b-scholz) [inline src/ram/MergeExtend.h:59]: Can we print "MERGE-EXTEND" with a hyphenation? In future, we may want to have a parser for RAM and we would like to have only a single keyword for the operation.

File: src/include/souffle/datastructure/EquivalenceRelation.h [hunk 3]
```diff
@@ -128,17 +128,28 @@ class EquivalenceRelation {
...
+     * new tuples" are now explicitly inserted this relation, and all of the new
+     * tuples in this relation are inserted into the old relation.
      */
-    void extend(const EquivalenceRelation<TupleType>& other) {
-        // nothing to extend if there's no new/original knowledge
-        if (other.size() == 0 || this->size() == 0) return;
+    void extendAndInsert(EquivalenceRelation<TupleType>& other) {
+        if (other.size() == 0 && this->size() == 0) return;
 
         std::set<value_type> repsCovered;
```
Comment (b-scholz) [inline src/include/souffle/datastructure/EquivalenceRelation.h:142]: Can we change this to an std::unordered_set?

File: src/include/souffle/datastructure/EquivalenceRelation.h [hunk 4]
```diff
@@ -128,17 +128,28 @@ class EquivalenceRelation {
...
+        // relation by other. These operations are interleaved for maximum
+        // efficiency - either extend or inserting first would make the other
+        // operation unnecessarily slow.
+        std::vector<std::pair<value_type, value_type>> toInsert;
+        auto size = std::distance(this->sds.sparseToDenseMap.begin(), this->sds.sparseToDenseMap.end());
+        toInsert.reserve(size);
+
         // find all the disjoint sets that need to be added to this relation
         // that exist in other (and exist in this)
         {
```
Comment (b-scholz) [inline src/include/souffle/datastructure/EquivalenceRelation.h:155]: Have you experimented swapping the traversals to compute `repsCovered`? 
You could traverse over `others` and find the containment in `this`.

Comment (b-scholz) [reply src/include/souffle/datastructure/EquivalenceRelation.h:155]: Just in case you would like to further tune the performance...

**Comment by b-scholz** [comment]
Another performance question: Langston discovered that extending the new relation is faster. We should experiment with this as well and adjust the code generation (i.e. the new knowledge becomes an over-approximation).

**Comment by kevzhumba** [comment]
> Can we print "MERGE-EXTEND" with a hyphenation? In future, we may want to have a parser for RAM and we would like to have only a single keyword for the operation.

Done

> Can we change this to an std::unordered_set?

Done

> Have you experimented swapping the traversals to compute repsCovered?
You could traverse over others and find the containment in this.

Tried this, did not see any performance improvement.

> Another performance question: Langston discovered that extending the new relation is faster. We should experiment with this as well and adjust the code generation (i.e. the new knowledge becomes an over-approximation).

I'm not sure I understand this comment. I thought the new knowledge, after extending, is already an over-approximation as described in the pact paper. 

> There is still the issue that the domain traversals of equivalence relations are too expensive at the moment...

I can take a look at this and post a new pull request once I have the time (end of the semester is always busy). 

Thanks for all the help!

**Comment by kevzhumba** [comment]
apologies for the formatting commits, forgot to run clang-format

**Comment by b-scholz** [comment]
There is this issue:
```
 230 |         return mk<ram::Sequence>(mk<ram::Extend>(destRelation, srcRelation), std::move(stmt));
185
      |                                          ^~~~~~
186
/home/runner/work/souffle/souffle/src/ast2ram/seminaive/UnitTranslator.cpp:230:75: error: no matching function for call to ‘mk<<expression error> >(const string&, const string&)’
187
  230 |         return mk<ram::Sequence>(mk<ram::Extend>(destRelation, srcRelation), std::move(stmt));
188
      |                                                                           ^
189
In file included from /home/runner/work/souffle/souffle/src/include/souffle/utility/DynamicCasting.h:20,
190
                 from /home/runner/work/souffle/souffle/src/include/souffle/utility/ContainerUtil.h:19,
191
                 from /home/runner/work/souffle/souffle/src/ast2ram/UnitTranslator.h:20,
192
                 from /home/runner/work/souffle/souffle/src/ast2ram/seminaive/UnitTranslator.h:17,
193
                 from /home/runner/work/souffle/souffle/src/ast2ram/seminaive/UnitTranslator.cpp:15:
194
/home/runner/work/souffle/souffle/src/include/souffle/utility/Types.h:57:8: note: candidate: ‘template<class A, class B, class ... Args> souffle::Own<A> souffle::mk(Args&& ...)’
195
   57 | Own<A> mk(Args&&... xs) {
196
      |        ^~
197
/home/runner/work/souffle/souffle/src/include/souffle/utility/Types.h:57:8: note:   template argument deduction/substitution failed:
198
/home/runner/work/souffle/souffle/src/ast2ram/seminaive/UnitTranslator.cpp:230:75: error: template argument 1 is invalid
199
  230 |         return mk<ram::Sequence>(mk<ram::Extend>(destRelation, srcRelation), std::move(stmt));
200
      |                                                                           ^
201
```
which needs to be fixed.

**Comment by kevzhumba** [comment]
Hi Bernhard,

I am unable to reproduce this failure; I am running 
```
cmake -S . -B build
cmake --build build
```
locally and after cleaning any old build files and am not getting any errors. Checking the code on this branch, I'm not able to find the line of code that is causing the error you pasted either. Langston's original commit had already removed that line of code.

**Comment by b-scholz** [comment]
Can you create a new PR to check whether this is an issue related to GitHub? 
It works for me as well when I pull your changes locally.

**Comment by b-scholz** [comment]
It seems that the wrong versions of files are picked (i.e. UnitTranslator.cpp).

**Pull request closed without merging**